### PR TITLE
[fix][DRAFT] buffer the fp16 grads

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
- fixes major issues in the "fp16 reduce" option which was on the repo (not what Mannat and I tested recently, I've been meaning to submit this PR for a while but got delayed)
- now the issue is that this implementation works unit test wise, works when testing locally, but hangs on FB cluster. I would love to have your eyes on that, I've had no success in trying to figure this out
cc @msbaines @anj-s @min-xu-ai 

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 1]
```diff
@@ -414,23 +416,27 @@ def reduce(*_: Any) -> None:
                     param.grad.mul_(self.world_size_scaling)
 
                     if self.reduce_fp16:
-                        param.grad.data = param.grad.data.half()
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:417]: this was wrong, param.grad is strongly typed and needs to be the same as params

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 2]
```diff
@@ -414,23 +416,27 @@ def reduce(*_: Any) -> None:
...
+                        self._grad_fp16_buffer[index] = param.grad.to(torch.float16)
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:419]: temporarily save a fp16 variant of the grad

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 3]
```diff
@@ -414,23 +416,27 @@ def reduce(*_: Any) -> None:
...
 
                     # Future work includes clearing up the buffer if possible
                     def cleanup() -> None:
                         if dst_rank != self.global_rank:
                             param.grad = None
-                        else:
-                            assert param.grad is not None
-                            param.grad.data = param.grad.data.to(dtype=param.dtype)
+                        elif self._grad_fp16_buffer[index] is not None:
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/sharded_ddp.py:425]: when the reduce is done, upgrade to fp32 and save in the proper place

File: tests/nn/data_parallel/test_pytorch_parity_sharded_ddp.py [hunk 4]
```diff
@@ -39,8 +39,12 @@
 
 _test_fp16_reduction = [False]
 
-if hasattr(dist, "algorithms.ddp_com_hooks.default_hooks"):
+try:
+    from torch.distributed.algorithms.ddp_comm_hooks.default_hooks import fp16_compress_hook
```
Comment (blefaudeux) [inline tests/nn/data_parallel/test_pytorch_parity_sharded_ddp.py:43]: tested, this works on recent pytorch

File: fairscale/nn/data_parallel/sharded_ddp.py [hunk 5]
```diff
@@ -414,23 +416,27 @@ def reduce(*_: Any) -> None:
...
+                        self._grad_fp16_buffer[index] = None
 
                     # Async reduce for this buffer, log the future
                     dst_global_rank = OSS.get_global_rank(self.process_group, dst_rank)
 
                     self._work_handles.append(
                         Workhandle(
                             handle=dist.reduce(
-                                tensor=param.grad.data, dst=dst_global_rank, group=self.process_group, async_op=True
+                                tensor=self._grad_fp16_buffer[index] if self.reduce_fp16 else param.grad,  # type: ignore  # mypy is drunk
```
Comment (anj-s) [inline fairscale/nn/data_parallel/sharded_ddp.py:436]: Should this be param.grad.data?

Comment (blefaudeux) [reply fairscale/nn/data_parallel/sharded_ddp.py]: I switched to that, AFAIK it should not change much though

**Comment by blefaudeux** [comment]
@anj-s example of FB run f255729495, the stack does not tell much. I'm now thinking that it may be an issue with inter stream sync, the fp16 cast is an op and it's in a different stream from nccl

**Comment by blefaudeux** [comment]
to be revived soon

**Pull request closed without merging**